src/zarr/core/metadata/dtype.py

```diff
@@ -0,0 +1,576 @@
+        case _:
+            return float(data)
+
+
+def float_from_json_v3(data: JSONFloat) -> float:
+    # todo: support the v3-specific NaN handling
+    return float_from_json_v2(data)
+
+
+def float_from_json(data: JSONFloat, zarr_format: ZarrFormat) -> float:
+    if zarr_format == 2:
+        return float_from_json_v2(data)
+    else:
+        return float_from_json_v3(data)
+
+
+def complex_from_json_v2(data: JSONFloat, dtype: Any) -> np.complexfloating:
+    return dtype.type(data)
+
+
+def complex_from_json_v3(data: tuple[JSONFloat, JSONFloat], dtype: Any) -> np.complexfloating:
+    return dtype.type(complex(*data))
+
+
+def complex_from_json(
+    data: tuple[JSONFloat, JSONFloat], dtype: Any, zarr_format: ZarrFormat
+) -> np.complexfloating:
+    if zarr_format == 2:
+        return complex_from_json_v2(data, dtype)
+    else:
+        if check_json_complex_float_v3(data):
+            return complex_from_json_v3(data, dtype)
+        else:
+            raise TypeError(f"Invalid type: {data}. Expected a sequence of two numbers.")
+    raise ValueError(f"Invalid zarr format: {zarr_format}. Expected 2 or 3.")
+
+
+TDType = TypeVar("TDType", bound=np.dtype[Any])
+TScalar = TypeVar("TScalar", bound=np.generic)
+
+
+class DTypeWrapper(Generic[TDType, TScalar], ABC, Metadata):
+    name: ClassVar[str]
+    dtype_cls: ClassVar[type[TDType]]  # this class will create a numpy dtype
+    kind: ClassVar[DataTypeFlavor]
+    default_value: TScalar
+
+    def __init_subclass__(cls) -> None:
+        # Subclasses will bind the first generic type parameter to an attribute of the class
+        # TODO: wrap this in some *very informative* error handling
+        generic_args = get_args(get_original_bases(cls)[0])
+        cls.dtype_cls = generic_args[0]
+        return super().__init_subclass__()
+
+    def to_dict(self) -> dict[str, JSON]:
+        return {"name": self.name}
+
+    def cast_value(self: Self, value: object, *, endianness: Endianness | None = None) -> TScalar:
+        return cast(np.generic, self.to_dtype(endianness=endianness).type(value))
+
+    @classmethod
+    @abstractmethod
+    def from_dtype(cls: type[Self], dtype: TDType) -> Self:
+        raise NotImplementedError
+
+    def to_dtype(self: Self, *, endianness: Endianness | None = None) -> TDType:
+        endian_str = endianness_to_numpy_str(endianness)
+        return self.dtype_cls().newbyteorder(endian_str)
+
+    @abstractmethod
+    def to_json_value(self, data: np.generic, *, zarr_format: ZarrFormat) -> JSON:
+        """
+        Convert a single value to JSON-serializable format. Depends on the zarr format.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def from_json_value(
+        self: Self, data: JSON, *, zarr_format: ZarrFormat, endianness: Endianness | None = None
+    ) -> TScalar:
+        """
+        Read a JSON-serializable value as a numpy scalar
+        """
+        raise NotImplementedError
+
```

Comment:
@nenb curious to get your thoughts on this API. One potentially confusing thing is that I use `__post_init__` to bind an instance of the generic type parameter `TDtype` to the `dtype_cls` class variable. This means we get type hints for the `to_dtype` method, AND we can actually create a dtype instance when we need to, all from the generic type parameter. I think this will need some very solid documentation though.

Reply:
oops, i meant to say that I'm using `__init_subclass__`, not `__post_init__`.